mapper-runtime/src/main/java/com/datastax/oss/driver/api/mapper/annotations/QueryProvider.java

```diff
@@ -0,0 +1,99 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.api.mapper.annotations;
+
+import com.datastax.oss.driver.api.mapper.MapperContext;
+import com.datastax.oss.driver.api.mapper.entity.EntityHelper;
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+/**
+ * Annotates a {@link Dao} method that delegates the execution of the query to a user-provided
+ * class.
+ *
+ * <p>Example:
+ *
+ * <pre>
+ * &#64;Dao
+ * public interface SensorDao {
+ *   &#64;QueryProvider(providerClass = FindSliceProvider.class, entityHelpers = SensorReading.class)
+ *   PagingIterable<SensorReading> findSlice(int id, Integer month, Integer day);
+ * }
+ *
+ * public class FindSliceProvider {
+ *   public FindSliceProvider(
+ *       MapperContext context, EntityHelper<SensorReading> sensorReadingHelper) {
+ *     ...
+ *   }
+ *
+ *   public PagingIterable<SensorReading> findSlice(int id, Integer month, Integer day) {
+ *     ... // implement the query logic here
+ *   }
+ * }
+ * </pre>
+ *
+ * <p>Use this (over {@link Query}) for requests that can't be expressed as static query strings,
+ * for example if some clauses are added dynamically depending on the values of some parameters.
+ *
+ * <p>The parameters and return type are completely free-form, as long as they match those of the
+ * provider method.
+ */
+@Target(ElementType.METHOD)
+@Retention(RetentionPolicy.CLASS)
+public @interface QueryProvider {
+
+  /**
+   * The class that will execute the query.
+   *
+   * <p>The mapper will create an instance of this class for each DAO instance. It must expose a
+   * constructor that is accessible from the DAO interface's package, and takes the following
+   * parameter types:
+   *
+   * <ul>
+   *   <li>{@link MapperContext}.
+   *   <li>zero or more {@link EntityHelper}s, as defined by {@link #entityHelpers()}.
+   * </ul>
+   *
+   * It must also expose a method that will be invoked each time the DAO method is called, see
+   * {@link #providerMethod()}.
+   */
+  Class<?> providerClass();
+
+  /**
+   * The method to invoke on the provider class.
+   *
+   * <p>It must be accessible from the DAO interface's package, and have the same return and
+   * parameter types as the annotated DAO method.
+   *
+   * <p>This is optional; if not provided, it defaults to the name of the annotated DAO method.
+   */
+  String providerMethod() default "";
+
+  /**
+   * A list of entities for which {@link EntityHelper} instances should be injected into the
+   * provider class's constructor (in addition to the mapper context).
+   *
+   * <p>For example, if {@code entityHelpers = {Product.class, Dimensions.class}}, your provider
```

Comment:
What are the use cases for specifying multiple `EntityHelper` classes?  The only thing that comes to mind is maybe batch statements?

Reply:
It could be useful for nested UDTs too.